File: pandas/general_query_12.py
#!/usr/bin/python3

"""General query 12

Student Mobility

Javier Fernández-Bravo Peñuela
Universitat Politècnica de València

Gender proportion of exchange students received by Spanish host institutions.

pandas - general_query_12.py
"""

from pandas_job.pandas_task import compute_task
from domain.student import Student, MobilityType
from util.functions import round_half_down

from sys import argv

SPAIN = "ES"

def visitor_gender_proportion(students_df):
    students_df['HOST_COUNTRY'] = students_df[['MOBILITYTYPE','COUNTRYOFHOSTINSTITUTION','COUNTRYOFWORKPLACEMENT']]\
        .apply(host_country_udf, axis=1)
    students_df = students_df.dropna(subset=['HOST_COUNTRY', 'GENDER'])
    visitor_students_df = students_df[students_df['HOST_COUNTRY'] == SPAIN].copy()
    visitor_students_df.loc[:, 'GENDER'] = visitor_students_df['GENDER'].apply(lambda g: g.value)
    student_count = visitor_students_df.shape[0]
    return visitor_students_df[['GENDER']].groupby(['GENDER']).size().sort_index()\
        .apply(lambda g: round_half_down(g / student_count))

def host_country_udf(row):
    host_country = ''
    if row['MOBILITYTYPE'] in [MobilityType.STUDY, MobilityType.PLACEMENT]:
        if row['MOBILITYTYPE'] is MobilityType.STUDY:
            host_country = row['COUNTRYOFHOSTINSTITUTION']
        elif row['MOBILITYTYPE'] is MobilityType.PLACEMENT:
            host_country = row['COUNTRYOFWORKPLACEMENT']
    return host_country.rstrip() if host_country else None

# python3 general_query_12.py ../data/Student_Mobility.csv [out/pandas12-output]
if __name__ == "__main__":
    compute_task(visitor_gender_proportion, Student, argv[1:])

File: spark/src/main/java/systematicchaos/studentmobility/GeneralQuery12.java
/**
 * Student Mobility
 * 
 * Javier Fernández-Bravo Peñuela
 * Universitat Politècnica de València
 * 
 * General query 12: Gender proportion of exchange students received by Spanish host institutions.
 * 
 * spark - GeneralQuery12.java
 */

package systematicchaos.studentmobility;

import static org.apache.spark.sql.functions.callUDF;
import static org.apache.spark.sql.functions.col;
import static systematicchaos.studentmobility.util.Functions.roundHalfDown;

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.api.java.UDF3;
import org.apache.spark.sql.types.DataTypes;

import systematicchaos.studentmobility.domain.Student;
import systematicchaos.studentmobility.spark.SparkJob;
import systematicchaos.studentmobility.spark.SparkJobLauncher;
import systematicchaos.studentmobility.spark.SparkTask;
import systematicchaos.studentmobility.util.KeyCountPair;
import systematicchaos.studentmobility.util.StudentBuilder;

public class GeneralQuery12 {
	
	public static class VisitorGenderProportion implements SparkTask<Double> {
		
		private static final String SPAIN = "ES";
		
		@Override
		public Iterable<KeyCountPair<Double>> computeSparkTask(String[] inputFiles) {
			SparkJob<Student> job = new SparkJob<>(GeneralQuery12.class.getName(),
					Student.class, new StudentBuilder());
			List<KeyCountPair<Double>> genderProportion = new ArrayList<>(2);
			
			try {
				Dataset<Student> students = job.readDataset(inputFiles);
				
				students.sqlContext().udf().register("HOST_COUNTRY",
						(UDF3<String, String, String, String>)(mobilityType, countryHostInstitution, countryWorkPlacement) -> {
							String hostCountry = null;
							if (mobilityType != null && !mobilityType.isEmpty()) {
								if (Student.MobilityType.STUDY.getValue() == mobilityType.charAt(0)) {
									hostCountry = countryHostInstitution;
								} else if (Student.MobilityType.PLACEMENT.getValue() == mobilityType.charAt(0)) {
									hostCountry = countryWorkPlacement;
								}
							}
							return hostCountry;
						}, DataTypes.StringType);
				
				Dataset<Row> visitorStudents = students.withColumn("hostCountry", callUDF("HOST_COUNTRY",
						col("mobilityType"), col("countryOfHostInstitution"), col("countryOfWorkPlacement")))
						.filter(col("hostCountry").equalTo(SPAIN));
				final double visitorStudentCount = (double) visitorStudents.count();
				Dataset<Row> visitorStudentsGender = visitorStudents.groupBy(col("gender")).count();
				
				visitorStudentsGender.orderBy(col("gender")).takeAsList(2).forEach(
						s -> genderProportion.add(new KeyCountPair<>(
								String.valueOf(s.getString(0).charAt(0)),
								roundHalfDown(s.getLong(1) / visitorStudentCount))));
			} catch (Exception e) {
				System.err.println(e.getClass().getName() + ": " + e.getMessage());
			} finally {
				job.close();
			}
			
			return genderProportion;
		}
	}
	
	// spark-submit --class systematicchaos.studentmobility.GeneralQuery12 --master local[*] target/spark-1.0.0.jar ../data/Student_Mobility.csv [out/spark12-output]
	public static void main(String[] args) throws Exception {
		SparkJobLauncher.launchSparkJob(new VisitorGenderProportion(), args);
	}
}
